File: src/window.cpp
#include <string.h>
#include <iostream>

#include "libnativeapi/include/nativeapi.h"
#include "window.h"

using namespace nativeapi;

extern WindowManager g_window_manager;

Window* __window(long window_id) {
  std::shared_ptr<Window> window_ptr = g_window_manager.Get(window_id);
  if (window_ptr != nullptr) {
    return window_ptr.get();
  }
  return nullptr;
}

FFI_PLUGIN_EXPORT
void window_focus(long window_id) {
  __window(window_id)->Focus();
}

FFI_PLUGIN_EXPORT
void window_blur(long window_id) {
  __window(window_id)->Blur();
}

FFI_PLUGIN_EXPORT
bool window_is_focused(long window_id) {
  return __window(window_id)->IsFocused();
}

FFI_PLUGIN_EXPORT
void window_show(long window_id) {
  __window(window_id)->Show();
}

FFI_PLUGIN_EXPORT
void window_show_inactive(long window_id) {
  __window(window_id)->ShowInactive();
}

FFI_PLUGIN_EXPORT
void window_hide(long window_id) {
  __window(window_id)->Hide();
}

FFI_PLUGIN_EXPORT
bool window_is_visible(long window_id) {
  return __window(window_id)->IsVisible();
}

FFI_PLUGIN_EXPORT
void window_maximize(long window_id) {
  __window(window_id)->Maximize();
}

FFI_PLUGIN_EXPORT
void window_unmaximize(long window_id) {
  __window(window_id)->Unmaximize();
}

FFI_PLUGIN_EXPORT
bool window_is_maximized(long window_id) {
  return __window(window_id)->IsMaximized();
}

FFI_PLUGIN_EXPORT
void window_minimize(long window_id) {
  __window(window_id)->Minimize();
}

FFI_PLUGIN_EXPORT
void window_restore(long window_id) {
  __window(window_id)->Restore();
}

FFI_PLUGIN_EXPORT
bool window_is_minimized(long window_id) {
  return __window(window_id)->IsMinimized();
}

FFI_PLUGIN_EXPORT
void window_set_full_screen(long window_id, bool is_full_screen) {
  __window(window_id)->SetFullScreen(is_full_screen);
}

FFI_PLUGIN_EXPORT
bool window_is_full_screen(long window_id) {
  return __window(window_id)->IsFullScreen();
}

// FFI_PLUGIN_EXPORT
// void _window_set_background_color(long window_id, NativeColor color);

// FFI_PLUGIN_EXPORT
// NativeColor window_get_background_color(long window_id);

FFI_PLUGIN_EXPORT
void window_set_bounds(long window_id, struct NativeRectangle bounds) {
  __window(window_id)->SetBounds(
      {bounds.x, bounds.y, bounds.width, bounds.height});
}

FFI_PLUGIN_EXPORT
struct NativeRectangle window_get_bounds(long window_id) {
  auto bounds = __window(window_id)->GetBounds();
  return {bounds.x, bounds.y, bounds.width, bounds.height};
}

FFI_PLUGIN_EXPORT
void window_set_size(long window_id, struct NativeSize size, bool animate) {
  __window(window_id)->SetSize({size.width, size.height}, animate);
}

FFI_PLUGIN_EXPORT
struct NativeSize window_get_size(long window_id) {
  auto size = __window(window_id)->GetSize();
  return {size.width, size.height};
}

FFI_PLUGIN_EXPORT
void window_set_content_size(long window_id, struct NativeSize size) {
  __window(window_id)->SetContentSize({size.width, size.height});
}

FFI_PLUGIN_EXPORT
NativeSize window_get_content_size(long window_id) {
  auto size = __window(window_id)->GetContentSize();
  return {size.width, size.height};
}

FFI_PLUGIN_EXPORT
void window_set_minimum_size(long window_id, struct NativeSize size) {
  __window(window_id)->SetMinimumSize({size.width, size.height});
}

FFI_PLUGIN_EXPORT
struct NativeSize window_get_minimum_size(long window_id) {
  auto size = __window(window_id)->GetMinimumSize();
  return {size.width, size.height};
}

FFI_PLUGIN_EXPORT
void window_set_maximum_size(long window_id, struct NativeSize size) {
  __window(window_id)->SetMaximumSize({size.width, size.height});
}

FFI_PLUGIN_EXPORT
struct NativeSize window_get_maximum_size(long window_id) {
  auto size = __window(window_id)->GetMaximumSize();
  return {size.width, size.height};
}

FFI_PLUGIN_EXPORT
void window_set_resizable(long window_id, bool is_resizable) {
  __window(window_id)->SetResizable(is_resizable);
}

FFI_PLUGIN_EXPORT
bool window_is_resizable(long window_id) {
  return __window(window_id)->IsResizable();
}

FFI_PLUGIN_EXPORT
void window_set_movable(long window_id, bool is_movable) {
  __window(window_id)->SetMovable(is_movable);
}

FFI_PLUGIN_EXPORT
bool window_is_movable(long window_id) {
  return __window(window_id)->IsMovable();
}

FFI_PLUGIN_EXPORT
void window_set_minimizable(long window_id, bool is_minimizable) {
  __window(window_id)->SetMinimizable(is_minimizable);
}

FFI_PLUGIN_EXPORT
bool window_is_minimizable(long window_id) {
  return __window(window_id)->IsMinimizable();
}

FFI_PLUGIN_EXPORT
void window_set_maximizable(long window_id, bool is_maximizable) {
  __window(window_id)->SetMaximizable(is_maximizable);
}

FFI_PLUGIN_EXPORT
bool window_is_maximizable(long window_id) {
  return __window(window_id)->IsMaximizable();
}

FFI_PLUGIN_EXPORT
void window_set_full_screenable(long window_id, bool is_full_screenable) {
  __window(window_id)->SetFullScreenable(is_full_screenable);
}

FFI_PLUGIN_EXPORT
bool window_is_full_screenable(long window_id) {
  return __window(window_id)->IsFullScreenable();
}

FFI_PLUGIN_EXPORT
void window_set_closable(long window_id, bool is_closable) {
  __window(window_id)->SetClosable(is_closable);
}

FFI_PLUGIN_EXPORT
bool window_is_closable(long window_id) {
  return __window(window_id)->IsClosable();
}

FFI_PLUGIN_EXPORT
void window_set_always_on_top(long window_id, bool is_always_on_top) {
  __window(window_id)->SetAlwaysOnTop(is_always_on_top);
}

FFI_PLUGIN_EXPORT
bool window_is_always_on_top(long window_id) {
  return __window(window_id)->IsAlwaysOnTop();
}

FFI_PLUGIN_EXPORT
void window_set_position(long window_id, struct NativePoint point) {
  __window(window_id)->SetPosition({point.x, point.y});
}

FFI_PLUGIN_EXPORT
struct NativePoint window_get_position(long window_id) {
  auto position = __window(window_id)->GetPosition();
  return {position.x, position.y};
}

FFI_PLUGIN_EXPORT
void window_set_title(long window_id, const char* title) {
  __window(window_id)->SetTitle(title);
}

FFI_PLUGIN_EXPORT
const char* window_get_title(long window_id) {
  static std::string title_storage = __window(window_id)->GetTitle();
  return title_storage.c_str();
}

FFI_PLUGIN_EXPORT
void window_set_has_shadow(long window_id, bool has_shadow) {
  __window(window_id)->SetHasShadow(has_shadow);
}

FFI_PLUGIN_EXPORT
bool window_has_shadow(long window_id) {
  return __window(window_id)->HasShadow();
}

FFI_PLUGIN_EXPORT
void window_set_opacity(long window_id, float opacity) {
  __window(window_id)->SetOpacity(opacity);
}

FFI_PLUGIN_EXPORT
float window_get_opacity(long window_id) {
  return __window(window_id)->GetOpacity();
}

FFI_PLUGIN_EXPORT
void window_set_visible_on_all_workspaces(long window_id,
                                          bool is_visible_on_all_workspaces) {
  __window(window_id)->SetVisibleOnAllWorkspaces(is_visible_on_all_workspaces);
}

FFI_PLUGIN_EXPORT
bool window_is_visible_on_all_workspaces(long window_id) {
  return __window(window_id)->IsVisibleOnAllWorkspaces();
}

FFI_PLUGIN_EXPORT
void window_set_ignore_mouse_events(long window_id,
                                    bool is_ignore_mouse_events) {
  __window(window_id)->SetIgnoreMouseEvents(is_ignore_mouse_events);
}

FFI_PLUGIN_EXPORT
bool window_is_ignore_mouse_events(long window_id) {
  return __window(window_id)->IsIgnoreMouseEvents();
}

FFI_PLUGIN_EXPORT
void window_set_focusable(long window_id, bool is_focusable) {
  __window(window_id)->SetFocusable(is_focusable);
}

FFI_PLUGIN_EXPORT
bool window_is_focusable(long window_id) {
  return __window(window_id)->IsFocusable();
}

FFI_PLUGIN_EXPORT
void window_start_dragging(long window_id) {
  __window(window_id)->StartDragging();
}

FFI_PLUGIN_EXPORT
void window_start_resizing(long window_id) {
  __window(window_id)->StartResizing();
}

FFI_PLUGIN_EXPORT
void window_start_listening(long window_id) {
  // TODO: Implement
}

FFI_PLUGIN_EXPORT
void window_stop_listening(long window_id) {
  // TODO: Implement
}

File: lib/src/window.dart
import 'dart:ui';
import 'dart:ffi' as ffi;

import 'package:ffi/ffi.dart';
import 'package:nativeapi/src/ffi/bindings.dart';
import 'package:nativeapi/src/ffi/bindings_generated.dart';

/// A class for representing a window.
///
/// This class provides methods for managing a window.
class Window {
  /// The native API bindings.
  NativeApiBindings get _bindings => nativeApiBindings;

  /// The ID of the window.
  final int id;

  /// Creates a new window.
  Window({required this.id});

  /// Focuses the window.
  void focus() {
    _bindings.window_focus(id);
  }

  /// Blurs the window.
  void blur() {
    _bindings.window_blur(id);
  }

  /// Checks if the window is focused.
  bool isFocused() {
    return _bindings.window_is_focused(id);
  }

  /// Shows the window.
  void show() {
    _bindings.window_show(id);
  }

  /// Hides the window.
  void hide() {
    _bindings.window_hide(id);
  }

  /// Checks if the window is visible.
  bool isVisible() {
    return _bindings.window_is_visible(id);
  }

  /// Maximizes the window.
  void maximize() {
    _bindings.window_maximize(id);
  }

  /// Unmaximizes the window.
  void unmaximize() {
    _bindings.window_unmaximize(id);
  }

  /// Checks if the window is maximized.
  bool isMaximized() {
    return _bindings.window_is_maximized(id);
  }

  /// Minimizes the window.
  void minimize() {
    _bindings.window_minimize(id);
  }

  /// Restores the window.
  void restore() {
    _bindings.window_restore(id);
  }

  /// Checks if the window is minimized.
  bool isMinimized() {
    return _bindings.window_is_minimized(id);
  }

  /// Sets the window to full screen.
  void setFullScreen(bool isFullScreen) {
    _bindings.window_set_full_screen(id, isFullScreen);
  }

  /// Checks if the window is full screen.
  bool isFullScreen() {
    return _bindings.window_is_full_screen(id);
  }

  // void SetBackgroundColor(Color color);
  // Color GetBackgroundColor() const;

  /// Sets the bounds of the window.
  void setBounds(Rect bounds) {}

  /// Gets the bounds of the window.
  Rect get bounds {
    final nativeRectangle = _bindings.window_get_bounds(id);
    return Rect.fromLTRB(
      nativeRectangle.x,
      nativeRectangle.y,
      nativeRectangle.x + nativeRectangle.width,
      nativeRectangle.y + nativeRectangle.height,
    );
  }

  /// Sets the size of the window.
  void setSize(Size size) {}

  /// Gets the size of the window.
  Size getSize() {
    final nativeSize = _bindings.window_get_size(id);
    return Size(nativeSize.width, nativeSize.height);
  }

  /// Sets the content size of the window.
  void setContentSize(Size size) {}

  /// Gets the content size of the window.
  Size getContentSize() {
    final nativeSize = _bindings.window_get_content_size(id);
    return Size(nativeSize.width, nativeSize.height);
  }

  /// Sets the minimum size of the window.
  void setMinimumSize(Size size) {}

  /// Gets the minimum size of the window.
  Size getMinimumSize() {
    final nativeSize = _bindings.window_get_minimum_size(id);
    return Size(nativeSize.width, nativeSize.height);
  }

  /// Sets the maximum size of the window.
  void setMaximumSize(Size size) {}

  /// Gets the maximum size of the window.
  Size getMaximumSize() {
    final nativeSize = _bindings.window_get_maximum_size(id);
    return Size(nativeSize.width, nativeSize.height);
  }

  /// Sets whether the window is resizable.
  void setResizable(bool isResizable) {
    _bindings.window_set_resizable(id, isResizable);
  }

  /// Checks if the window is resizable.
  bool isResizable() {
    return _bindings.window_is_resizable(id);
  }

  /// Sets whether the window is movable.
  void setMovable(bool isMovable) {
    _bindings.window_set_movable(id, isMovable);
  }

  /// Checks if the window is movable.
  bool isMovable() {
    return _bindings.window_is_movable(id);
  }

  /// Sets whether the window is minimizable.
  void setMinimizable(bool isMinimizable) {
    _bindings.window_set_minimizable(id, isMinimizable);
  }

  /// Checks if the window is minimizable.
  bool isMinimizable() {
    return _bindings.window_is_minimizable(id);
  }

  /// Sets whether the window is maximizable.
  void setMaximizable(bool isMaximizable) {
    _bindings.window_set_maximizable(id, isMaximizable);
  }

  /// Checks if the window is maximizable.
  bool isMaximizable() {
    return _bindings.window_is_maximizable(id);
  }

  /// Sets whether the window is full screenable.
  void setFullScreenable(bool isFullScreenable) {
    _bindings.window_set_full_screenable(id, isFullScreenable);
  }

  /// Checks if the window is full screenable.
  bool isFullScreenable() {
    return _bindings.window_is_full_screenable(id);
  }

  /// Sets whether the window is closable.
  void setClosable(bool isClosable) {
    _bindings.window_set_closable(id, isClosable);
  }

  /// Checks if the window is closable.
  bool isClosable() {
    return _bindings.window_is_closable(id);
  }

  /// Sets whether the window is always on top.
  void setAlwaysOnTop(bool isAlwaysOnTop) {
    _bindings.window_set_always_on_top(id, isAlwaysOnTop);
  }

  /// Checks if the window is always on top.
  bool isAlwaysOnTop() {
    return _bindings.window_is_always_on_top(id);
  }

  /// Sets the position of the window.
  void setPosition(Offset offset) {
    final pointPtr = malloc<NativePoint>();
    pointPtr.ref.x = offset.dx;
    pointPtr.ref.y = offset.dy;
    _bindings.window_set_position(id, pointPtr.ref);
  }

  /// Gets the position of the window.
  Offset getPosition() {
    final nativePoint = _bindings.window_get_position(id);
    return Offset(nativePoint.x, nativePoint.y);
  }

  /// Sets the title of the window.
  void setTitle(String title) {}

  /// Gets the title of the window.
  String getTitle() {
    return _bindings.window_get_title(id).cast<Utf8>().toDartString();
  }

  /// Sets whether the window has a shadow.
  void setHasShadow(bool hasShadow) {
    _bindings.window_set_has_shadow(id, hasShadow);
  }

  /// Checks if the window has a shadow.
  bool hasShadow() {
    return _bindings.window_has_shadow(id);
  }

  /// Sets the opacity of the window.
  void setOpacity(double opacity) {
    _bindings.window_set_opacity(id, opacity);
  }

  /// Gets the opacity of the window.
  double getOpacity() {
    return _bindings.window_get_opacity(id);
  }

  /// Sets whether the window is focusable.
  void setFocusable(bool isFocusable) {
    _bindings.window_set_focusable(id, isFocusable);
  }

  /// Checks if the window is focusable.
  bool isFocusable() {
    return _bindings.window_is_focusable(id);
  }

  /// Starts dragging the window.
  void startDragging() {
    _bindings.window_start_dragging(id);
  }

  /// Starts resizing the window.
  void startResizing() {
    _bindings.window_start_resizing(id);
  }

  void addListener(VoidCallback listener) {}

  @override
  int get hashCode => id.hashCode;

  @override
  operator ==(Object other) {
    return other is Window && other.id == id;
  }
}
